src/legacy/core_plugins/embeddable_api/public/embeddables/embeddable.tsx

```diff
@@ -0,0 +1,185 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { isEqual, cloneDeep } from 'lodash';
+import { Adapters } from 'ui/inspector';
+import * as Rx from 'rxjs';
+import { IContainer } from '../containers';
+import { IEmbeddable, EmbeddableInput, EmbeddableOutput } from './i_embeddable';
+import { ViewMode } from '../types';
+
+function getPanelTitle({
+  inputTitle,
+  defaultTitle,
+  hidePanelTitles,
+}: {
+  inputTitle?: string;
+  defaultTitle?: string;
+  hidePanelTitles?: boolean;
+}) {
+  if (hidePanelTitles) {
+    return '';
+  }
+
+  // Specificaly check for undefined, as an empty string override means "hide this title".
+  return inputTitle === undefined ? defaultTitle : inputTitle;
+}
+
+export abstract class Embeddable<
+  TEmbeddableInput extends EmbeddableInput = EmbeddableInput,
+  TEmbeddableOutput extends EmbeddableOutput = EmbeddableOutput
+> implements IEmbeddable<TEmbeddableInput, TEmbeddableOutput> {
+  public readonly parent?: IContainer;
+  public readonly isContainer: boolean = false;
+  public abstract readonly type: string;
+  public readonly id: string;
+
+  protected output: TEmbeddableOutput;
+  protected input: TEmbeddableInput;
+
+  private readonly input$: Rx.BehaviorSubject<TEmbeddableInput>;
+  private readonly output$: Rx.BehaviorSubject<TEmbeddableOutput>;
+
+  // Listener to parent changes, if this embeddable exists in a parent, in order
+  // to update input when the parent changes.
+  private parentSubscription?: Rx.Subscription;
+
+  private destoyed: boolean = false;
+
+  constructor(input: TEmbeddableInput, output: TEmbeddableOutput, parent?: IContainer) {
+    this.id = input.id;
+    this.output = {
+      title: getPanelTitle({
+        inputTitle: input.title,
+        defaultTitle: output.defaultTitle,
```

Comment:
Yea, output doesn't change over the course of the lifetime of the embeddable, while input can.  That's why I put defaultTitle on output instead of input.